locator/tablets.cc

```diff
@@ -84,17 +84,24 @@ tablet_migration_streaming_info get_migration_streaming_info(const tablet_info&
 }
 
 tablet_migration_streaming_info get_migration_streaming_info(const tablet_info& tinfo, const tablet_transition_info& trinfo) {
-    tablet_migration_streaming_info result = {
-        .read_from = std::unordered_set<tablet_replica>(tinfo.replicas.begin(), tinfo.replicas.end()),
-        .written_to = std::unordered_set<tablet_replica>(trinfo.next.begin(), trinfo.next.end())
-    };
-    for (auto&& r : trinfo.next) {
-        result.read_from.erase(r);
-    }
-    for (auto&& r : tinfo.replicas) {
-        result.written_to.erase(r);
+    tablet_migration_streaming_info result;
+    switch (trinfo.transition) {
+        case tablet_transition_kind::migration:
+            result.read_from = std::unordered_set<tablet_replica>(tinfo.replicas.begin(), tinfo.replicas.end());
+            result.written_to = std::unordered_set<tablet_replica>(trinfo.next.begin(), trinfo.next.end());
+            for (auto&& r : trinfo.next) {
+                result.read_from.erase(r);
+            }
+            for (auto&& r : tinfo.replicas) {
+                result.written_to.erase(r);
+            }
+            return result;
+        case tablet_transition_kind::rebuild:
+            result.written_to = std::unordered_set<tablet_replica>(trinfo.next.begin(), trinfo.next.end());
+            result.read_from = result.written_to;
+            return result;
```

Comment:
I don't understand this at all. Don't we want to stream to just the new tablet replica? And stream from all tablets, excluding the one on the dead nodes, and the new tablet?

Reply:
Note that this is for the purpose of load impact tracking only.

This is anticipating that we will use repair, which may write to other replicas too.

Reply:
Maybe we should have separate fields for that. I really liked the consolidation where this function identifies sources and targets, and then the rest of the code doesn't care about the transition kind. Not critical for merging, though, but comments are needed to explain the unexpected computation.